test/jdk/java/foreign/critical/TestCritical.java

```diff
@@ -0,0 +1,177 @@
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @library ../ /test/lib
+ *
+ * @run testng/othervm --enable-native-access=ALL-UNNAMED TestCritical
+ */
+
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.FunctionDescriptor;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentAllocator;
+import java.lang.foreign.SequenceLayout;
+import java.lang.foreign.StructLayout;
+import java.lang.foreign.ValueLayout;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.VarHandle;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.IntFunction;
+import java.util.stream.Stream;
+
+import static org.testng.Assert.assertEquals;
+
+public class TestCritical extends NativeTestHelper {
+
+    static {
+        System.loadLibrary("Critical");
+    }
+
+    @Test
+    public void testEmpty() throws Throwable {
+        MethodHandle handle = downcallHandle("empty", FunctionDescriptor.ofVoid(), Linker.Option.critical(false));
+        handle.invokeExact();
+    }
+
+    @Test
+    public void testIdentity() throws Throwable {
+        MethodHandle handle = downcallHandle("identity", FunctionDescriptor.of(C_INT, C_INT), Linker.Option.critical(false));
+        int result = (int) handle.invokeExact(42);
+        assertEquals(result, 42);
+    }
+
+    @Test
+    public void testWithReturnBuffer() throws Throwable {
+        StructLayout bigLayout = MemoryLayout.structLayout(
+                C_LONG_LONG.withName("x"),
+                C_LONG_LONG.withName("y"));
+
+        MethodHandle handle = downcallHandle("with_return_buffer", FunctionDescriptor.of(bigLayout), Linker.Option.critical(false));
+        VarHandle vhX = bigLayout.varHandle(MemoryLayout.PathElement.groupElement("x"));
+        VarHandle vhY = bigLayout.varHandle(MemoryLayout.PathElement.groupElement("y"));
+        try (Arena arena = Arena.ofConfined()) {
+            MemorySegment result = (MemorySegment) handle.invokeExact((SegmentAllocator) arena);
+            long x = (long) vhX.get(result, 0L);
+            assertEquals(x, 10);
+            long y = (long) vhY.get(result, 0L);
+            assertEquals(y, 11);
+        }
+    }
+
+    public record AllowHeapCase(IntFunction<MemorySegment> newArraySegment, ValueLayout elementLayout,
+                                String fName, FunctionDescriptor fDesc) {}
+
+    @Test(dataProvider = "allowHeapCases")
+    public void testAllowHeap(AllowHeapCase testCase) throws Throwable {
```

Comment:
Note that this is the only new test. The diff looks bigger because I've renamed the enclosing directory.